File: go/parser.go
package parsejoy
import "crypto/sha1"
import "encoding/hex"
import "fmt"
import "strings"
import "7scientists.com/parsejoy/set"

type State interface{
    Copy() State
    //Pushes a new AST node onto the AST stack
    PushASTNode(node *ASTNode, state State) error
    //Pushes an AST property onto the AST stack
    PushASTProperty(property *ASTProperty, state State) error
}

/*
$ast-node: Calls PushASTNode with a new node
$ast-property: Calls PushASTProperty with a new property

PushASTNode should check which values are currently on the stack.
It should go through each of these values and make sure that it's
and ASTProperty. If it is, it should get incorporated into the
current ASTNode.

PushASTProperty should just push the property onto the current
stack.
*/

type ASTNode struct{
    //represents an AST node
    Type string
    Attributes map[string]interface{}
}

func (node *ASTNode) Initialize(nodeType string, properties map[interface{}]interface{}) {
    node.Type = nodeType
    node.Attributes = make(map[string]interface{})
}

type ASTProperty struct{
    //represents an AST property
    Name string
    List bool
    Value interface{}
}

type BaseToken interface{
    SetNext(token BaseToken)
    GetValue(state State) string
}

type CompilerError struct{
    msg string
}

type ParserError struct{
    msg string
}

func (e *CompilerError) Error() string {return e.msg}
func (e *ParserError) Error() string {return e.msg}

type Parser func(State) (State, BaseToken, error)
type ParserMap map[string]Parser

type ParserPlugin interface {
    setPrefixes(name string, prefix interface{})
    PrepareGrammar(grammar map[string]interface{})
    //getPrefixSet(state State) set.Set
    resolveRule(pg *ParserGenerator, name string) ((func (rule interface{}) (Parser, error) ) , bool)
    parser(parser Parser,name string,rule interface{},emit bool) Parser
    getPrefix(rule interface{},args interface{}) set.Set
    getEmptyPrefix(args interface{}) set.Set
    SetGenerator(generator *ParserGenerator)
}

type ParserGenerator struct {
    grammar map[string]interface{}
    plugin ParserPlugin
    parsers ParserMap
}

type ParserPluginMixin struct {
    RulePrefixes map[string]interface{}
    generator *ParserGenerator
    fingerprintIds map[string]uint
    fingerprintId uint
}

type OutcomeMixin struct {
    Outcome [][]Outcome
    OutcomeMap []uint
}

func PrintASTNode(node *ASTNode, level int) int {
	nodes := 1
	fmt.Println(strings.Repeat(" ", level*2), node.Type,":")
    for key, value := range node.Attributes {
        astNodeValue, ok := value.(*ASTNode)
        if ok {
            nodes += PrintASTTree(astNodeValue, level*2)
            continue
        }
        astNodeList, ok := value.([]interface{})
        if ok {
            fmt.Println(strings.Repeat(" ", level*2+1), key,":")
            for i, item := range astNodeList {
                astNodeItem, ok := item.(*ASTNode)
                if ok {
                    nodes += PrintASTTree(astNodeItem, level+1)
                } else {
                    fmt.Println(strings.Repeat(" ", level*2+1), key, "(", i, ")=", astNodeItem)
                }
            }
            continue
        }
        fmt.Println(strings.Repeat(" ", level*2+1), key, "=", value)
    }
	return nodes
}

func PrintASTTree(node interface{}, level int) int {
    astNode, ok := node.(*ASTNode)
    if ok {
        return PrintASTNode(astNode, level)
    }
    nodes := 1
    astNodeList, ok := node.([]interface{})
    if ok {
        for _, item := range astNodeList {
            fmt.Println(item)
            nodes += PrintASTTree(item, level)
        }
    }
    return nodes
}


func (om *OutcomeMixin) Initialize(N uint,M uint) {
    om.OutcomeMap = make([]uint,N)
    om.Outcome = make([][]Outcome,0,M)
}

func (pg *ParserPluginMixin) setPrefixes(name string, prefix interface{}) {
    pg.RulePrefixes[name] = prefix
}

func (pg *ParserPluginMixin) Initialize() {
    pg.fingerprintIds = make(map[string]uint)
    pg.RulePrefixes = make(map[string]interface{})
    pg.fingerprintId = 1
}

func (pg *ParserPluginMixin) SetGenerator (generator *ParserGenerator){
    pg.generator = generator
}

func (pg *ParserGenerator) SetPlugin(plugin ParserPlugin) {
    pg.plugin = plugin
    plugin.SetGenerator(pg)
}

func (pg *ParserGenerator) GetFingerprint (rule interface{}) string {

    data := "foobar"
    stringRule, ok := rule.(string)
    if ok {
        data = data+stringRule
    } else {
        listRule, ok := rule.([]interface{})
        if ok {
            for i := range listRule {
                element := listRule[i]
                data = data+pg.GetFingerprint(element)
            }
        } else {
            mapRule, ok := rule.(map[interface{}]interface{})

            if ok {
                for key, value := range mapRule {
                    data = data+pg.GetFingerprint(key)
                    data = data+pg.GetFingerprint(value)
                }
            }
        }
    }

    byteData := []byte(data)
    s1 := sha1.Sum(byteData)
    return hex.EncodeToString(s1[:])
}

func (pg *ParserGenerator) getPrefixes (rule interface{}, visitedRules set.HashSet, args interface{}, onlyStart bool) set.Set {

    stringRule, ok := rule.(string)

    if ok {
        if visitedRules.Contains(stringRule) {
            return pg.plugin.getEmptyPrefix(args)
        }
        visitedRules.Add(stringRule)
        grammarRule, ok := pg.grammar[stringRule]
        if ok {
            prefixes := pg.getPrefixes(grammarRule,visitedRules,args, onlyStart)
            pg.plugin.setPrefixes(stringRule, prefixes)
            return prefixes
        }
        return pg.plugin.getPrefix(stringRule,args)
    }

    listRule, ok := rule.([]interface{})

    if ok && len(listRule) > 0{
        prefixes := pg.plugin.getEmptyPrefix(args)
        for i:= range listRule  {
            ruleItem := listRule[i]
            subrulePrefixes := pg.getPrefixes(ruleItem, visitedRules,args,onlyStart)
            //if the current prefixes contain a false value (i.e. an optional token), we first remove it
            //before adding the new prefixes
            if prefixes.Contains(false) {
                _ = prefixes.Remove(false)
            }
            prefixes, _ = prefixes.Union(subrulePrefixes)
            //if we match only the start and this contains no false value, we break from the loop
            if onlyStart && !prefixes.Contains(false){
                break
            }
        }
        return prefixes
    }

    dictRule, ok := rule.(map[interface{}]interface{})

    if ok {
        if len(dictRule) == 1 {
            var key, value interface{}
            for key, value = range dictRule {
                break
            }
            stringKey, ok := key.(string)
            if !ok {
                panic("This should not happen!")
            }
            switch stringKey {
                case "$or":
                    listValue, ok := value.([]interface{})
                    if !ok || len(listValue) == 0 {
                        panic("This should not happen!")
                    }
                    prefixes := pg.plugin.getEmptyPrefix(args)
                    for i := range listValue {
                        subrule := listValue[i]
                        newPrefixes := pg.getPrefixes(subrule, visitedRules, args, onlyStart)
                        prefixes, _ = prefixes.Union(newPrefixes)
                    }
                    return prefixes
                case "$ast-prop": fallthrough
                case "$ast-node":
                    dictValue, ok := value.(map[interface{}]interface{})
                    if !ok {
                    }
                    return pg.getPrefixes(dictValue["value"], visitedRules, args, onlyStart)
                case "$and": fallthrough
                case "$repeat":
                    return pg.getPrefixes(value, visitedRules, args, onlyStart)
                case "$not":
                    nilPrefix := pg.plugin.getEmptyPrefix(args)
                    nilPrefix.Add(nil)
                    return nilPrefix
                case "$optional":
                    newPrefixes := pg.getPrefixes(value, visitedRules,args, onlyStart)
                    newPrefixes.Add(false)
                    return newPrefixes
                default:
                    return pg.plugin.getPrefix(rule,args)
            }
        }
    }
    panic("this should never happen")
}

func (pg *ParserGenerator) compileSequence (rules []interface{}) (Parser, error) {

    parsers := make([]Parser,len(rules))

    for i, rule := range rules {
        parser, ok := pg.compileRule(rule)
        if (ok != nil){
            return nil, ok
        }
        parsers[i] = parser
    }

    sequenceParser := func(state State) (State, BaseToken, error) {
        currentState := state
        var err error
        var baseToken BaseToken
        var currentToken BaseToken
        for i := range parsers{
            var newToken BaseToken
            currentState, newToken, err = parsers[i](currentState)
            if newToken != nil {
                if baseToken == nil {
                    baseToken = newToken
                }
                if currentToken != nil {
                    currentToken.SetNext(newToken)
                }
                currentToken = newToken
            }
            if err != nil{
                if baseToken != nil{
                    baseToken.SetNext(nil)
                }
                return state, nil, err
            }
        }
        return currentState, baseToken, nil
    }

    return pg.parser(sequenceParser, "seq", rules, false), nil
}

func (pg *ParserGenerator) parser(fn Parser, name string, rule interface{}, emit bool) Parser {

    if pg.plugin != nil {
        return pg.plugin.parser(fn,name,rule,emit)
    }
    wrappedParser := func(state State) (State, BaseToken, error){
        return fn(state)
    }
    return wrappedParser
}

func (pg *ParserGenerator) extractSubrule(rule interface{}, name string) (interface{}, bool) {
    ruleDict, ok := rule.(map[interface{}]interface{})

    if !ok {
        return nil, false
    }

    resolvedRule, ok := ruleDict[name]

    return resolvedRule, ok

}

func (pg *ParserGenerator) compileRepeat (rule interface{}) (Parser, error) {

    repeatRule, ok := pg.extractSubrule(rule, "$repeat")

    if !ok{
        return nil, &CompilerError{"repeat rule not found!"}
    }

    parser, err := pg.compileRule(repeatRule)

    if err != nil{
        return nil, err
    }

    pe := &ParserError{"repeat did not match once!"}

    repeatParser := func(state State) (State, BaseToken, error) {
        cnt := 0
        currentState := state
        var err error
        var baseToken BaseToken
        var currentToken BaseToken
        for {
            var newToken BaseToken
            currentState, newToken, err = parser(currentState)
            if newToken != nil {
                if baseToken == nil {
                    baseToken = newToken
                }
                if currentToken != nil {
                    currentToken.SetNext(newToken)
                }
                currentToken = newToken
            }
            if err != nil {
                break
            }
            cnt += 1
        }
        if cnt == 0 {
            if baseToken != nil{
                baseToken.SetNext(nil)
            }
            return state, nil, pe
        }
        return currentState, baseToken, nil
    }
    return pg.parser(repeatParser, "$repeat", rule, false), nil
}


func (pg *ParserGenerator) compileOr (rule interface{}) (Parser, error) {

    orRule, ok := pg.extractSubrule(rule, "$or")

    ruleArray, ok := orRule.([]interface{})

    if !ok{
        return nil, &CompilerError{"Expected a list of rules!"}
    }

    alternatives := make([]Parser,len(ruleArray))


    for i := range ruleArray {
        alternativeRule := ruleArray[i]
        compiledRule, err := pg.compileRule(alternativeRule)
        if err != nil{
            return nil, err
        }
        alternatives[i] = compiledRule
    }

    pe := &ParserError{"No alternative matched!"}


    orParser := func(state State) (State, BaseToken, error) {
        for i := range alternatives {
            newState, baseToken, err := alternatives[i](state)
            if err == nil{
                return newState, baseToken, nil
            }
        }
        return state, nil, pe
    }
    return pg.parser(orParser, "or", rule, false), nil

}


func (pg *ParserGenerator) compileAnd (rule interface{}) (Parser, error) {

    andRule, ok := pg.extractSubrule(rule, "$and")

    if !ok{
        return nil, &CompilerError{"and rule not found!"}
    }

    parser, err := pg.compileRule(andRule)

    if err != nil{
        return nil, err
    }

    pe := &ParserError{"And condition did not match!"}

    andParser := func(state State) (State, BaseToken, error) {
        stateCopy := state.Copy()
        _, _, err := parser(stateCopy)
        if err == nil {
            return state, nil, nil
        }
        return state, nil, pe
    }
    return pg.parser(andParser, "and", rule, false), nil

}

func (pg *ParserGenerator) compileNot (rule interface{}) (Parser, error) {

    notRule, ok := pg.extractSubrule(rule, "$not")

    if !ok{
        return nil, &CompilerError{"not rule not found!"}
    }

    parser, err := pg.compileRule(notRule)

    if err != nil{
        return nil, err
    }

    notParser := func(state State) (State, BaseToken, error) {
        stateCopy := state.Copy()
        _, _, err := parser(stateCopy)
        if err != nil {
            return state, nil, nil
        }
        return state, nil, &ParserError{"Not condition did match!"}
    }
    return pg.parser(notParser, "not", rule, false), nil

}

func (pg *ParserGenerator) compileOptional (rule interface{}) (Parser, error) {

    optionalRule, ok := pg.extractSubrule(rule, "$optional")

    if !ok{
        return nil, &CompilerError{"optional rule not found!"}
    }

    parser, err := pg.compileRule(optionalRule)

    if err != nil{
        return nil, err
    }

    optionalParser := func(state State) (State, BaseToken, error) {
        newState, baseToken, err := parser(state)
        if err != nil {
            return state, nil, nil
        }
        return newState, baseToken, nil
    }
    return pg.parser(optionalParser, "optional", rule, false), nil

}

func (pg *ParserGenerator) compileASTProperty(rule interface{}) (Parser, error) {

  astRule, ok := pg.extractSubrule(rule, "$ast-prop")

  if !ok{
      return nil, &CompilerError{"AST property rule not found!"}
  }

  astStringRule, ok := astRule.(map[interface{}]interface{})

  if !ok {
    return nil, &CompilerError{"Expected a map"}
  }

  valueRule, ok := astStringRule["value"]

  if !ok {
    return nil, &CompilerError{"You need to specify a value!"}
  }

  name, ok := astStringRule["name"]

  if !ok {
      return nil, &CompilerError{"You need to specify a name!"}
  }

  stringName, ok := name.(string)

  if !ok {
      return nil, &CompilerError{"Name must be a string!"}
  }

  //properties is an optional field
  asList, ok := astStringRule["as-list"]

  var asListBool bool

  if ok {
      asListBool, ok = asList.(bool)
      if !ok {
          return nil, &CompilerError{"as-list must be a boolean!"}
      }
  }

  asLiteral, ok := astStringRule["as-literal"]

  var asLiteralBool bool

  if ok {
      asLiteralBool, ok = asLiteral.(bool)
      if !ok {
          return nil, &CompilerError{"as-literal must be boolean!"}
      }
  }

  parser, err := pg.compileRule(valueRule)

  if err != nil{
      return nil, err
  }

  /*
  * Get the length of the current AST stack
  * When the parser is finished, move all new objects in the
    current AST stack into the AST property and then
    push the property itself onto the current stack.
  */
  astParser := func(state State) (State, BaseToken, error) {
      newState, baseToken, err := parser(state)
      if err == nil {
          astProperty := new(ASTProperty)
          astProperty.Name = stringName
          astProperty.List = asListBool
          if asLiteralBool {
              astProperty.Value = baseToken.GetValue(newState)
          }
          newState.PushASTProperty(astProperty, state)
      }
      return newState, baseToken, err
  }

  return astParser,nil

}

func (pg *ParserGenerator) compileASTNode(rule interface{}) (Parser, error) {

  astRule, ok := pg.extractSubrule(rule, "$ast-node")

  if !ok{
      return nil, &CompilerError{"AST-node rule not found!"}
  }

  astStringRule, ok := astRule.(map[interface{}]interface{})

  if !ok {
    return nil, &CompilerError{"Expected a map"}
  }

  valueRule, ok := astStringRule["value"]

  if !ok {
    return nil, &CompilerError{"You need to specify a value!"}
  }

  nodeType, ok := astStringRule["type"]

  if !ok {
      return nil, &CompilerError{"You need to specify an node type!"}
  }

  stringNodeType, ok := nodeType.(string)

  if !ok {
      return nil, &CompilerError{"Node type must be a string!"}
  }

  //properties is an optional field
  properties, ok := astStringRule["properties"]

  var mapProperties map[interface{}]interface{}

  if ok {
      mapProperties, ok = properties.(map[interface{}]interface{})
      if !ok {
          return nil, &CompilerError{"Properties must be a map!"}
      }
  }

  parser, err := pg.compileRule(valueRule)

  if err != nil{
      return nil, err
  }

  astParser := func(state State) (State, BaseToken, error) {
      newState, baseToken, err := parser(state)
      if err == nil {
          astNode := new(ASTNode)
          astNode.Initialize(stringNodeType, mapProperties)
          newState.PushASTNode(astNode, state)
      }
      return newState, baseToken, err
  }

  return astParser,nil

}

func (pg *ParserGenerator) resolveRule(name string) ((func (rule interface{}) (Parser, error) ) , bool) {
    switch name {
    case "$ast-node":
        return pg.compileASTNode, true
    case "$ast-prop":
        return pg.compileASTProperty, true
    case "$repeat":
        return pg.compileRepeat, true
    case "$or":
        return pg.compileOr, true
    case "$and":
        return pg.compileAnd, true
    case "$not":
        return pg.compileNot, true
    case "$optional":
        return pg.compileOptional, true
    default:
        if pg.plugin != nil {
            return pg.plugin.resolveRule(pg, name)
        }
    }
    return nil, false
}

func (pg *ParserGenerator) compileRule (rule interface{}) (Parser, error) {
    name := ""
    if rn, ok := rule.(string);ok {
        name = rn
        if parser, okr := pg.parsers[name]; okr {
            return parser, nil
        } else {
            if grammarRule, okr := pg.grammar[name]; okr {
                rule = grammarRule
            } else {

                if generator, ok := pg.resolveRule(name); ok {
                    return generator(rule)
                }
                return nil, &CompilerError{"Unknown rule: "+rn}
            }
        }
    }

    parseSubrule := func(rule map[interface{}]interface{}, name string) (Parser, error) {
        if len(rule) == 1 {
            var ruleName string
            var ok bool
            for key, _ := range rule{
                ruleName, ok = key.(string)
                break
            }
            if !ok{
                return nil,&CompilerError{"Key is not a string!"}
            }
            if generator, ok := pg.resolveRule(ruleName); ok {
                parser, err := generator(rule)
                if err != nil{
                    return nil, err
                }
                if name != "" {
                    pg.parsers[name] = pg.parser(parser, name, rule, true)
                    return pg.parsers[name], nil
                }
                return parser, nil
            } else{
                return nil, &CompilerError{"Unknown rule!"}
            }
        }
        return nil, nil
    }

    if name != "" {
        subruleParser := func(state State) (State, BaseToken, error) {
            return pg.parsers[name](state)
        }
        pg.parsers[name] = subruleParser
    }
    switch rt := rule.(type) {
    case map[interface{}]interface{}:
        if len(rt) == 1 {
            return parseSubrule(rt, name)
        }

    case []interface{}:
        if sequenceParser, err := pg.compileSequence(rt); err != nil{
            return nil, err
        } else {
            if name != "" {
                pg.parsers[name] = pg.parser(sequenceParser, name, rule, true)
                return pg.parsers[name], nil
            }
            return sequenceParser, nil
        }
    case string:
        if subrule, err := pg.compileRule(rt); err != nil{
            return nil, err
        } else{
            wrappedParser := pg.parser(subrule, name, rule, true)
            pg.parsers[name] = wrappedParser
            return wrappedParser, nil
        }
    default:
        return nil, &CompilerError{"Unknown rule type"}

    }

    return nil, &CompilerError{"Unknown rule encountered!"}

}


func (pg *ParserGenerator) Compile (grammar map[string]interface{}) (Parser, error) {

    pg.grammar = grammar
    pg.parsers = make(ParserMap)

    if pg.plugin == nil {
        return nil, &CompilerError{"No plugin defined!"}
    }

    pg.plugin.PrepareGrammar(grammar)

    if  _, ok := grammar["start"]; !ok{
        panic("Grammar does not contain a start rule!")
    }

    return pg.compileRule(grammar["start"])


}

File: cpp/src/parser.cc
#include "grammar.h"
#include "parser.h"
#include "sha.h"
#include <iostream>
#include <string>

using namespace std;

std::string string_to_hex(const std::string& input)
{
    static const char* const lut = "0123456789ABCDEF";
    size_t len = input.length();

    std::string output;
    output.reserve(2 * len);
    for (size_t i = 0; i < len; ++i)
    {
        const unsigned char c = input[i];
        output.push_back(lut[c >> 4]);
        output.push_back(lut[c & 15]);
    }
    return output;
}

namespace sscientists {
namespace parsejoy {

State::~State() {
}

State::State(LuaEnvironment& luaEnvironment) : luaEnvironment_(luaEnvironment) {
}

State::State(const State& state) : luaEnvironment_(state.luaEnvironment_) {
}

unsigned int State::Advance(const unsigned int n){
    pos_ += n;
    return pos_;
}

unsigned int State::Position(){
    return pos_;
}

void State::SetPosition(unsigned int n){
    pos_ = n;
}

ParserGenerator::ParserGenerator(const shared_ptr<Rule> g, LuaEnvironment& env) : luaEnvironment_(env) {
    debug = false;
    assert(g->type() == Type::MAP);
    grammar = static_pointer_cast<MapRule>(g);
};

std::string ParserGenerator::getFingerprint(const shared_ptr<Rule> rule){

    CryptoPP::SHA256 sha256;

    const std::string value = "foo";

    sha256.Update((const byte *)value.c_str(),value.size());

    switch(rule->type()){
        case Type::STRING:{
            auto stringRule = static_pointer_cast<StringRule>(rule);
            sha256.Update((const byte *)stringRule->rule.c_str(),stringRule->rule.size());
            break;
        }
        case Type::SEQUENCE:{
            auto sequenceRule = static_pointer_cast<SequenceRule>(rule);
            for(auto it=sequenceRule->rules.begin();it!=sequenceRule->rules.end();++it){
                std::string result = getFingerprint(*it);
                sha256.Update((const byte *)result.c_str(),result.size());
            }
            break;
        }
        case Type::MAP:{
            auto mapRule = static_pointer_cast<MapRule>(rule);
            for(auto it=mapRule->rules.begin();it!=mapRule->rules.end();++it){
                std::string key = it->first;
                sha256.Update((const byte *)key.c_str(),key.size());
                std::string result = getFingerprint(it->second);
                sha256.Update((const byte *)result.c_str(),result.size());
            }
            break;
        }
        default:{
            break;
        }
    }

    byte* bytes = new byte[sha256.DigestSize()];
    sha256.Final(bytes);
    auto s = reinterpret_cast<const char *>(bytes);
    return std::string(s);
}

//Set& ParserGenerator::getPrefixes(const shared_ptr<Rule> rule,HashSet& visitedRules,void *args,bool onlyStart){
    /*
    stringRule, ok := rule.(string)

    if ok {
        if visitedRules.Contains(stringRule) {
            return pg.plugin.getEmptyPrefix(args)
        }
        visitedRules.Add(stringRule)
        grammarRule, ok := pg.grammar[stringRule]
        if ok {
            prefixes := pg.getPrefixes(grammarRule,visitedRules,args, onlyStart)
            pg.plugin.setPrefixes(stringRule, prefixes)
            return prefixes
        }
        return pg.plugin.getPrefix(stringRule,args)
    }

    listRule, ok := rule.([]interface{})

    if ok && len(listRule) > 0{
        prefixes := pg.plugin.getEmptyPrefix(args)
        for i:= range listRule  {
            ruleItem := listRule[i]
            subrulePrefixes := pg.getPrefixes(ruleItem, visitedRules,args,onlyStart)
            //if the current prefixes contain a false value (i.e. an optional token), we first remove it
            //before adding the new prefixes
            if prefixes.Contains(false) {
                _ = prefixes.Remove(false)
            }
            prefixes, _ = prefixes.Union(subrulePrefixes)
            //if we match only the start and this contains no false value, we break from the loop
            if onlyStart && !prefixes.Contains(false){
                break
            }
        }
        return prefixes
    }

    dictRule, ok := rule.(map[interface{}]interface{})

    if ok {
        if len(dictRule) == 1 {
            var key, value interface{}
            for key, value = range dictRule {
                break
            }
            stringKey, ok := key.(string)
            if !ok {
                panic("This should not happen!")
            }
            switch stringKey {
                case "$or":
                    listValue, ok := value.([]interface{})
                    if !ok || len(listValue) == 0 {
                        panic("This should not happen!")
                    }
                    prefixes := pg.plugin.getEmptyPrefix(args)
                    for i := range listValue {
                        subrule := listValue[i]
                        newPrefixes := pg.getPrefixes(subrule, visitedRules, args, onlyStart)
                        prefixes, _ = prefixes.Union(newPrefixes)
                    }
                    return prefixes
                case "$ast-prop": fallthrough
                case "$ast-node":
                    dictValue, ok := value.(map[interface{}]interface{})
                    if !ok {
                    }
                    return pg.getPrefixes(dictValue["value"], visitedRules, args, onlyStart)
                case "$and": fallthrough
                case "$repeat":
                    return pg.getPrefixes(value, visitedRules, args, onlyStart)
                case "$not":
                    nilPrefix := pg.plugin.getEmptyPrefix(args)
                    nilPrefix.Add(nil)
                    return nilPrefix
                case "$optional":
                    newPrefixes := pg.getPrefixes(value, visitedRules,args, onlyStart)
                    newPrefixes.Add(false)
                    return newPrefixes
                default:
                    return pg.plugin.getPrefix(rule,args)
            }
        }
    }
    panic("this should never happen")
    */
//}

Parser ParserGenerator::compileLua(const shared_ptr<Rule> rule, RuleCache& cache){

    const shared_ptr<Rule> subRule = extractSubrule(rule, "$lua");
    auto stringRule = dynamic_pointer_cast<StringRule>(subRule);

    auto compiledLua = luaEnvironment_.compileCode(stringRule->rule);

    auto luaParser = [compiledLua, this](State& state){
        unsigned int pos = state.Position();
        //we place the state as a global variable
        //we can actually do this only once while the parser is running...
        this->luaEnvironment_.place("state", (void *)&state,state.className(), false);
        auto result = this->luaEnvironment_.runByteCode(compiledLua);
        auto error = result != 0;
        if (error)
            state.SetPosition(pos);
        return ParserResult(emptyToken, error);
    };

    return luaParser;

}


Parser ParserGenerator::compileSequence(const shared_ptr<Rule> rule, RuleCache& cache){

    std::vector<Parser> parsers;
    auto sequenceRule = dynamic_pointer_cast<SequenceRule>(rule);

    for(auto it=sequenceRule->rules.begin();it!=sequenceRule->rules.end();++it){
        auto parser = wrap(compileRule(*it, "", cache),"$sequence", rule, false);
        parsers.push_back(parser);
    }

    auto sequenceParser = [parsers](State& state){
        unsigned int pos = state.Position();
        for(auto it=parsers.begin();it!=parsers.end();++it){
            auto result = (*it)(state);
            bool error = std::get<1>(result);
            if (error == true){
                state.SetPosition(pos);
                return ParserResult(emptyToken, true);
            }
        }
        return ParserResult(emptyToken, false);
    };

    return sequenceParser;

}

Parser ParserGenerator::wrap(Parser parser, std::string name, const shared_ptr<Rule> rule, bool emit){
    auto wrapper = [name, parser, this](State& state){
        cout << &state << "\n";
        auto result = parser(state);
        return result;
    };
    return wrapper;
}

const shared_ptr<Rule> ParserGenerator::extractSubrule(const shared_ptr<Rule> rule, const std::string name){
    /*
    We might not need this function anymore...
    */
    try{
        auto mapRule = dynamic_pointer_cast<MapRule>(rule);
        auto it = mapRule->rules.find(name);
        if (it == mapRule->rules.end())
            throw std::runtime_error("Expected a "+name+" entry!");
        return it->second;
    } catch (const std::bad_cast& e){
        throw std::runtime_error("Expected a map rule in extractSubrule!");
    }
}

Parser ParserGenerator::compileRepeat(const shared_ptr<Rule> rule, RuleCache& cache){

    const shared_ptr<Rule> subRule = extractSubrule(rule, "$repeat");
    auto subruleParser = compileRule(subRule, "", cache);

    auto repeatParser = [subruleParser](State& state){
        bool success = false;
        unsigned int pos = state.Position();
        while(true){
            auto result = subruleParser(state);
            bool error = std::get<1>(result);
            if (error){
                if (success)
                    return ParserResult(emptyToken, false);
                state.SetPosition(pos);
                return result;
            }
            success = true;
        }
    };

    return repeatParser;
}

Parser ParserGenerator::compileOr(const shared_ptr<Rule> rule,RuleCache& cache){

    const shared_ptr<Rule> subRule = extractSubrule(rule, "$or");
    std::vector<Parser> parsers;
    auto sequenceRule = dynamic_pointer_cast<SequenceRule>(subRule);

    for(auto it=sequenceRule->rules.begin();it!=sequenceRule->rules.end();++it){
        auto parser = compileRule(*it, "", cache);
        parsers.push_back(parser);
    }

    auto orParser = [this,parsers](State& state){
        for(auto it=parsers.begin();it!=parsers.end();++it){
            unsigned int pos = state.Position();
            auto result = (*it)(state);
            bool error = std::get<1>(result);
            if (!error)
                return result;
            state.SetPosition(pos);
            }
        return ParserResult(emptyToken, true);
    };

    return orParser;

}

Parser ParserGenerator::compileAnd(const shared_ptr<Rule> rule, RuleCache& cache){

    const shared_ptr<Rule> subRule = extractSubrule(rule, "$and");
    auto subruleParser = compileRule(subRule, "", cache);

    auto andParser = [subruleParser](State& state){
        unsigned int pos = state.Position();
        auto result = subruleParser(state);
        state.SetPosition(pos);
        bool error = std::get<1>(result);
        if (error)
            return ParserResult(emptyToken, true);
        return ParserResult(emptyToken, false);
    };

    return andParser;
}

Parser ParserGenerator::compileNot(const shared_ptr<Rule> rule, RuleCache& cache){

    const shared_ptr<Rule> subRule = extractSubrule(rule, "$not");
    auto subruleParser = compileRule(subRule, "", cache);

    auto notParser = [subruleParser](State& state){
        unsigned int pos = state.Position();
        auto result = subruleParser(state);
        state.SetPosition(pos);
        bool error = std::get<1>(result);
        if (error)
            return ParserResult(emptyToken, false);
        return ParserResult(emptyToken, true);
    };

    return notParser;
}

Parser ParserGenerator::compileOptional(const shared_ptr<Rule> rule, RuleCache& cache){

    const shared_ptr<Rule> subRule = extractSubrule(rule, "$optional");
    auto subruleParser = compileRule(subRule, "", cache);

    auto optionalParser = [subruleParser](State& state){
        unsigned int pos = state.Position();
        auto result = subruleParser(state);
        bool error = std::get<1>(result);
        if (error){
            state.SetPosition(pos);
            return ParserResult(emptyToken, false);
        }
        return result;
    };
    return optionalParser;
}

ruleParser ParserGenerator::resolveRule(const std::string name){
    cout << "Resolving rule |" << name << "|\n";
    if (name == "$lua")
        return std::bind(&ParserGenerator::compileLua, this, placeholders::_1, placeholders::_2);
    if (name == "$repeat")
        return std::bind(&ParserGenerator::compileRepeat, this, placeholders::_1, placeholders::_2);
    if (name == "$or")
        return std::bind(&ParserGenerator::compileOr, this, placeholders::_1, placeholders::_2);
    if (name == "$and")
        return std::bind(&ParserGenerator::compileAnd, this, placeholders::_1, placeholders::_2);
    if (name == "$not")
        return std::bind(&ParserGenerator::compileNot, this, placeholders::_1, placeholders::_2);
    if (name == "$optional")
        return std::bind(&ParserGenerator::compileOptional, this, placeholders::_1, placeholders::_2);
    throw std::runtime_error("Could not resolve rule "+name+"!");
}

Parser ParserGenerator::compileRule(const shared_ptr<Rule> rule, const std::string name, RuleCache& cache){

    /*
        Compiles a given rule.
    */

    if (name != "") {
        //we check if we have the rule in the cache. If yes, we return it from there...
        if (cache.find(name) != cache.end())
            return cache.find(name)->second;
        auto subruleParser = [name, &cache](State& state) {
            if (cache.find(name) == cache.end())
                throw std::runtime_error("Unable to find parser for rule "+name);
            return cache[name](state);
        };
        if (debug)
            cout << "Setting cache function for name " << name << "\n";
        cache[name] = subruleParser;
    }

    auto parseSubrule = [&](const shared_ptr<MapRule> rule, std::string subruleName){
        if (rule->rules.size() == 1){
            std::string ruleName = rule->rules.begin()->first;
            const shared_ptr<Rule> subRule = rule->rules.begin()->second;
            auto generator = resolveRule(ruleName);
            try{
                return wrap(generator(rule, cache), ruleName, rule, false);
            }catch(const std::runtime_error& e){
                cerr << "Error when parsing rule " << ruleName << " within context of " << name << "\n";
                throw;
            }

        } else {
            throw std::runtime_error("expected exactly one rule in subrule "+subruleName+" when parsing rule "+name);
        }
    };

    Parser parser;

    switch (rule->type()){
        case Type::STRING:{
            /*
            This is just a string, referencing another rule from the grammar.
            */

            auto newName = static_pointer_cast<StringRule>(rule)->rule;
            auto it = parserMap.find(name);
            if (it != parserMap.end()){
                parser = it->second;
            } else {
                auto grammarRule = grammar->rules.find(newName);
                if (grammarRule != grammar->rules.end()){
                    parser = compileRule(grammarRule->second, newName, cache);
                } else {
                    auto generator = resolveRule(newName);
                    parser = generator(rule, cache);
                }
            }
            break;
        }
        case Type::MAP:{
            auto mapRule = static_pointer_cast<MapRule>(rule);
            parser = parseSubrule(mapRule, name);
            break;
        }
        case Type::SEQUENCE:{
            parser = wrap(compileSequence(rule, cache), name, rule, true);
            break;
        }
    }

    if (name != ""){
        cache[name] = parser;
    }

    return parser;

}

Parser ParserGenerator::compile(){
    return compile("start");
}

Parser ParserGenerator::compile(const std::string& startRule){
    auto parser = compileRule(grammar->rules[startRule],startRule, cache);
    return parser;
}


}
}
